moshi/src/main/java/com/squareup/moshi/Types.java

```diff
@@ -137,6 +141,38 @@ static Type canonicalize(Type type) {
     }
   }
 
+  public static <T extends Annotation> T createJsonQualifierImplementation(
+      final Class<T> annotationType) {
+    if (!annotationType.isAnnotationPresent(JsonQualifier.class)) {
+      throw new IllegalArgumentException(annotationType + " does not have @JsonQualifier.");
+    }
+    if (annotationType.getDeclaredMethods().length != 0) {
+      throw new IllegalArgumentException(annotationType + " must not declare methods.");
+    }
+    // noinspection unchecked
+    return (T) Proxy.newProxyInstance(annotationType.getClassLoader(),
+        new Class<?>[] { annotationType }, new InvocationHandler() {
+          @Override public Object invoke(Object proxy, Method method, Object[] args)
+              throws Throwable {
+            String methodName = method.getName();
+            if (methodName.equals("annotationType")) {
+              return annotationType;
+            }
+            if (methodName.equals("equals")) {
+              Object o = args[0];
+              return o.getClass().isInstance(annotationType);
+            }
+            if (methodName.equals("hashCode")) {
+              return 0;
+            }
+            if (methodName.equals("toString")) {
+              return "@" + annotationType.getName() + "()";
+            }
+            throw new AssertionError();
```

Comment:
I think you probably want just `return method.invoke(this, args)`. Or at least that’s what I’ve done elsewhere.